### DISCUSSION TITLE
Breaking out OBX Segment into Multiple OBX segments

### DISCUSSION THREAD

**Discussion by noMoreUserNamesAvailableAnymore** [nextgenhealthcare/connect#4746] [Q&A] [answered]
Also posted to the older mirthproject.io website:
I am needing to take the info in OBX.5 and create a new OBX segment for every tilde ~ the rest of the message would be left alone.

Make:
`OBX|1|TX|TESTCODE^TEST NAME^L^LOINC^TEST DESCRIPTION^||blah~blah blah~Blah~blu blah blah|||N|||Z||||||||||||^^^^^^^ABC^^123`

Into:
```
OBX|1|TX|TESTCODE^TEST NAME^L^LOINC^TEST DESCRIPTION^||blah|||N|||Z||||||||||||^^^^^^^ABC^^ 123
OBX|2|TX|TESTCODE^TEST NAME^L^LOINC^TEST DESCRIPTION^||blah blah|||N|||Z||||||||||||^^^^^^^ABC^^123
OBX|3|TX|TESTCODE^TEST NAME^L^LOINC^TEST DESCRIPTION^||Blah|||N|||Z||||||||||||^^^^^^^ABC^^ 123
OBX|4|TX|TESTCODE^TEST NAME^L^LOINC^TEST DESCRIPTION^||blu blah blah|||N|||Z||||||||||||^^^^^^^ABC^^123
```

I previously used iteration to break this down for a report writer but also had to lean on functionality for that to "kinda" succeed there, but as the evolution of the status quo goes, I now have to also do this for a channel writer destination.

**Comment by pacmano1** [comment]
Do note you posted this on the forums.   Same answer.  Give it a try.

**Reply by pacmano1** [reply]
Poster used https://forums.mirthproject.io/forum/mirth-connect/support/7226-spliting-a-obx-record

**Comment by noMoreUserNamesAvailableAnymore** [comment, marked as answer]
Answer was proved by jkrebs when posted this fantastic code here: [https://forums.mirthproject.io/forum/mirth-connect/support/7226-spliting-a-obx-record?p=178718#post178718)](https://forums.mirthproject.io/forum/mirth-connect/support/7226-spliting-a-obx-record?p=178718#post178718)

But since there is a terrible habit on the internet to have links break eventually here are the contents of that post.  Again thanks jkrebs where ever you are.

05-06-2021, 01:14 PM
The below will traverse OBX groups for repeating OBX.5, create an OBX segment for all but the first iteration, populate OBX.5 in the new OBX, delete all but the first OBX.5 from originating OBX, insert the new OBXs then re-sequence the set IDs.
Code:
```
for each (obr in msg.OBR) {
  for each (obx in getSegmentsAfter(msg,obr,'OBX',false)) {
     if (obx['OBX.5'].length() > 1) {
        for (var x = obx['OBX.5'].length()-1; x > 0; x--) {
           var newObx = obx.copy()
           newObx['OBX.5'] = obx['OBX.5'][x]
           delete obx['OBX.5'][x]
           msg.insertChildAfter(obx, newObx)
        }
     }
  }
  var setId = 1
  for each (obx in getSegmentsAfter(msg,obr,'OBX',false))
    obx['OBX.1'] = setId++
}
```
Note that it assumes data in the components for which it creates OBXs. For a link to the getSegmentsAfter code template 
[https://forums.mirthproject.io/forum/mirth-connect/support/12373-getsegmentsafter](https://forums.mirthproject.io/forum/mirth-connect/support/12373-getsegmentsafter)
Last edited by jkrebs; 05-07-2021, 06:23 AM. Reason: verbiage
